### PR TITLE
Tool to ensure consistency between algorithms, unit definitions and glossary

### PR DESCRIPTION
**Pull request by tbody-cfs**
An idea for a tool that could be helpful.

We need to keep the `default_units` and `physics_glossary` up to date with the algorithm keys. We can try to do that manually, but it's a bit tedious. Instead, I implemented the following helper tool to make sure that these always stay consistent.

If this seems like a good idea, I'll polish it and add the units/glossary entries for everything there.

### CODE REVIEW

File: cfspopcon/default_units.yaml [hunk 1]
```diff
@@ -1,173 +1,174 @@
...
 core_radiator: null
-critical_alpha_MHD: ""
-current_relaxation_time: "s"
-cylindrical_safety_factor: ""
+core_radiator_charge_state: ''
+core_radiator_concentration: ''
+critical_alpha_MHD: ''
+current_relaxation_time: s
+cylindrical_safety_factor: ''
+definition: UNKNOWN
```
Comment (hassec) [inline cfspopcon/default_units.yaml:27]: What's `definition`?

File: cfspopcon/key_checker.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
+"""A simple tool to ensure consistency between the algorithms, default units and the physics glossary, with a CLI."""
```
Comment (hassec) [inline cfspopcon/key_checker.py:1]: I think we should move this script outside `cfspopcon` as a helper script which we can then run in CI to ensure things won't break. 

We could put it into a `tests/utils/` folder for example.

File: cfspopcon/key_checker.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        issue_detected = False
+
+        issue_detected = self.fix_unit_dictionary(dry_run) or issue_detected
+        issue_detected = self.fix_glossary(dry_run) or issue_detected
+
+        return issue_detected
+
+    def read_physics_glossary(self) -> tuple[dict[str, list[str]], set[str]]:
+        """Read the physics glossary file."""
+        with as_file(files("cfspopcon").parents[0] / "docs" / "doc_sources" / "physics_glossary.rst") as filepath:  # type:ignore[attr-defined]
```
Comment (hassec) [inline cfspopcon/key_checker.py:35]: If we move the script you can just assume that the script is run from the repo, and won't need to use `as_file` patterns.

File: cfspopcon/key_checker.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+        return glossary, set(glossary.keys())
+
+    def read_algorithm_keys(self) -> set[str]:
+        """Read the algorithm keys."""
+        algorithm_keys = []
+
+        for alg in Algorithm.instances.values():
+            for key in alg.input_keys + alg.return_keys:
+                if key not in algorithm_keys:
+                    algorithm_keys.append(key)
```
Comment (hassec) [inline cfspopcon/key_checker.py:96]: What do you think about directly making it a set and using `update` ? 
```suggestion
        algorithm_keys: set[str] = {}

        for alg in Algorithm.instances.values():
            algorithm_keys.update(alg.input_keys)
            algorithm_keys.update(alg.return_keys)
```

File: cfspopcon/default_units.yaml (hunk 1 above)
Comment (tbody-cfs) [reply cfspopcon/default_units.yaml]: I've renamed this to something clearer

File: tests/utils/variable_consistency_checker.py [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+        new_variables_dict = dict()
+        for key in sorted((self.variable_keys - unused_variable_keys) | unlisted_args, key=str.lower):
+            used_by = []
+            set_by = []
+
+            for alg in Algorithm.instances.values():
+                if key in alg.input_keys:
+                    used_by.append(alg._name)
+                if key in alg.return_keys:
+                    set_by.append(alg._name)
```
Comment (hassec) [inline tests/utils/variable_consistency_checker.py:145]: could consider moving this out of the loop and traversing the algorithms list once to create a dict of key -> (used_by, set_by). Would avoid having to do this full loop for every key.

**Review by hassec — CHANGES_REQUESTED** [review]
Minor set of comments, but excited to have this 👍

File: tests/utils/variable_consistency_checker.py [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+        for i, line in enumerate(glossary_text):
+            if ":sorted:" in line:
+                header_line = i
+                break
+        if header_line == 0:
+            raise RuntimeError("Header not found")
+        self._glossary_header = glossary_text[: header_line + 1]
+
+        # Define patterns for identifying blank lines, keys and descriptions.
+        pattern_for_key = re.compile(r"^\s{2}[^\s]")
```
[the remark this answers is not in the document]
Comment (tbody-cfs) [reply tests/utils/variable_consistency_checker.py]: Made change, and also changed `pattern_for_description`.
```
pattern_for_key = re.compile(r"^\s{2}\S+\s*$")
pattern_for_description = re.compile(r"^\s{4}\S.+")
```

**Review by hassec — APPROVED** [review]

**Pull request merged**